### PR TITLE
[TEST ONLY] Fix soundwire interrupt hardening

### PR DESCRIPTION
**Pull request by plbossart**
Compile-tested ideas to try to harden interrupt handling and clock restart sequences.

@bardliao @RanderWang can you take a look?

### CODE REVIEW

File: drivers/soundwire/cadence_master.c [hunk 1]
```diff
@@ -86,10 +87,12 @@ MODULE_PARM_DESC(cdns_mcp_int_mask, "Cadence MCP IntMask");
 #define CDNS_MCP_INT_DATA_CLASH			BIT(9)
 #define CDNS_MCP_INT_PARITY			BIT(8)
 #define CDNS_MCP_INT_CMD_ERR			BIT(7)
+#define CDNS_MCP_INT_RESERVED2			GENMASK(6, 4)
 #define CDNS_MCP_INT_RX_NE			BIT(3)
 #define CDNS_MCP_INT_RX_WL			BIT(2)
 #define CDNS_MCP_INT_TXE			BIT(1)
 #define CDNS_MCP_INT_TXF			BIT(0)
+#define CDNS_MCP_INT_RESERVED (CDNS_MCP_INT_RESERVED1 & CDNS_MCP_INT_RESERVED2)
```
Comment (bardliao) [inline drivers/soundwire/cadence_master.c]: Should this be (CDNS_MCP_INT_RESERVED1 | CDNS_MCP_INT_RESERVED2)?

File: drivers/soundwire/intel.c [hunk 2]
```diff
@@ -1589,15 +1589,16 @@ static int intel_suspend_runtime(struct device *dev)
 		intel_shim_wake(sdw, false);
 	} else if (clock_stop_quirks & SDW_INTEL_CLK_STOP_BUS_RESET ||
 		!clock_stop_quirks) {
-		ret = sdw_cdns_clock_stop(cdns, true);
+
+		ret = sdw_cdns_enable_interrupt(cdns, false);
```
Comment (bardliao) [inline drivers/soundwire/intel.c]: There are sdw transmission in the sdw_cdns_clock_stop() function, and interrupt is required. So we can't disable interrupt before calling sdw_cdns_clock_stop().
```
sdw_cdns_clock_stop() -> sdw_bus_prep_clk_stop() -> sdw_slave_clk_stop_callback() 
                                                 -> sdw_slave_clk_stop_prepare()
```

File: drivers/soundwire/cadence_master.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/cadence_master.c]: right. late-evening programming is never good.

File: drivers/soundwire/intel.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/intel.c]: ok, that's right

**Comment by plbossart** [comment]
@bardliao @RanderWang updated version pushed.

We have a race condition when doing a bus reset, we should wait for the devices to be re-enumerated before doing a clock stop deprepare. For now I disabled this case, but it cannot possibly work reliably without timeouts.

I still see the "Program transport params failed: -22" error, it's likely that when doing a reset we don't clear some registers or fields.

**Comment by plbossart** [comment]
@YvonneYang2 could you retest with the updated branch (commit 48956dfca545) and check if we have any 'Invalid interrupt' logs?

I would to understand if the interrupt filtering is actually necessary, and in what sequence this can happen. I have a hunch that making sure the states are correct is enough.

**Comment by YvonneYang2** [comment]
Tested it with latest plbossart:fix/soundwire-interrupt-hardening branch (commit: https://github.com/thesofproject/linux/commit/48956dfca545639052e650ca5259e6dee8eaf236), Run suspend/resume stress test cannot reproduced #1638 . But run multipipline stress test can still reproduce issue like https://github.com/thesofproject/linux/issues/1637#issuecomment-569228295 said.  Failed at 222nd time.
```
rt715 sdw:3:25d:715:0: Initialization not complete, timed out
```

Logs: 
[failed_multipipeline_test_222_all_logger.log](https://github.com/thesofproject/linux/files/4018313/failed_multipipeline_test_222_all_logger.log)
[failed_multipipeline_test_222_dmesg.log](https://github.com/thesofproject/linux/files/4018314/failed_multipipeline_test_222_dmesg.log)

"Invalid interrupt" occurred in each iteration of multipipline or suspend/resume stress test. 
Logs:
[multipipeline_test_1_all_logger.log](https://github.com/thesofproject/linux/files/4018318/multipipeline_test_1_all_logger.log)
[multipipeline_test_1_dmesg.log](https://github.com/thesofproject/linux/files/4018319/multipipeline_test_1_dmesg.log)

[test_1.log](https://github.com/thesofproject/linux/files/4018353/test_1.log)

**Comment by plbossart** [comment]
@YvonneYang2 thanks for the tests, would you mind retrying with the updated commit d5727490522ccf031c969be6dc1c5f58545a216d?
I think the error logs might be misleading, we want to know if there is an invalid interrupt or just an interrupt but with a zero status (nothing to do).

**Comment by plbossart** [comment]
> @YvonneYang2 thanks for the tests, would you mind retrying with the updated commit [d572749](https://github.com/thesofproject/linux/commit/d5727490522ccf031c969be6dc1c5f58545a216d)?
> I think the error logs might be misleading, we want to know if there is an invalid interrupt or just an interrupt but with a zero status (nothing to do).

Commit 88d8c0db8595 rather, the previous one had a compilation issue...

**Comment by YvonneYang2** [comment]
Tested it with latest fix/soundwire-interrupt-hardening (88d8c0d) , https://github.com/thesofproject/linux/issues/1638 can be reproduced now. And can also reproduce issue like https://github.com/thesofproject/linux/issues/1637#issuecomment-569228295 said. 

Invalid interrupt log is still in each iteration of suspend/resume and multipipline stress test, and all of it is 
```
sdw_cdns_irq: Invalid interrupt, nothing to do with status 0
```

Suspend/resume stress test log:
[s3_test_467_logger_all.log](https://github.com/thesofproject/linux/files/4024853/s3_test_467_logger_all.log)
[test_467.log](https://github.com/thesofproject/linux/files/4024854/test_467.log)

Multipipline stress test log:
[failed_multipipeline_test_26_all_logger.log](https://github.com/thesofproject/linux/files/4024855/failed_multipipeline_test_26_all_logger.log)
[failed_multipipeline_test_26_dmesg.log](https://github.com/thesofproject/linux/files/4024856/failed_multipipeline_test_26_dmesg.log)

File: drivers/soundwire/cadence_master.c [hunk 3]
```diff
@@ -1375,6 +1375,12 @@ int sdw_cdns_clock_restart(struct sdw_cdns *cdns, bool bus_reset)
 
 	cdns_update_config(cdns);
 
+	ret = sdw_cdns_enable_interrupt(cdns, true);
```
Comment (RanderWang) [inline drivers/soundwire/cadence_master.c:1398]: @plbossart  the issue “rt715 sdw:3:25d:715:0: Initialization not complete, timed out” should be caused by this change.
the normal status should be
````
[  207.838359] intel-sdw sdw-master-3: intel_resume_runtime start
[  207.838361] intel-sdw sdw-master-3: intel_resume_runtime resume from BUS_RESET
[  207.838974] rt715 sdw:3:25d:715:0: sdw_modify_slave_status: changing status slave 1 status 1 new status 0
[  207.838975] rt715 sdw:3:25d:715:0: sdw_modify_slave_status: initializing completion for Slave 1
[  207.838976] intel-sdw sdw-master-3: sdw_cdns_clock_restart: start
[  207.839217] intel-sdw sdw-master-3: sdw_cdns_clock_restart: end
[  207.839218] intel-sdw sdw-master-3: intel_resume_runtime done
[  207.840113] intel-sdw sdw-master-3: Slave status change
[  207.840118] intel-sdw sdw-master-3: Slave 0 ATTACHED
[  207.840119] intel-sdw sdw-master-3: sdw_handle_slave_status: start
[  207.840120] intel-sdw sdw-master-3: Slave attached, programming device number
[  207.840295] intel-sdw sdw-master-3: SDW Slave Addr: 21025d071500
[  207.840296] intel-sdw sdw-master-3: SDW Slave class_id 0, part_id 715, mfg_id 25d, unique_id 1, version 2
````
The abnormal case is:
````
[  218.431430] intel-sdw sdw-master-3: intel_resume_runtime start
[  218.431434] intel-sdw sdw-master-3: intel_resume_runtime resume from BUS_RESET
[  218.432009] rt715 sdw:3:25d:715:0: sdw_modify_slave_status: changing status slave 1 status 1 new status 0
[  218.432011] rt715 sdw:3:25d:715:0: sdw_modify_slave_status: initializing completion for Slave 1
[  218.432015] intel-sdw sdw-master-3: sdw_cdns_clock_restart: start
[  218.432036] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  218.432088] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  218.434677] intel-sdw sdw-master-3: sdw_cdns_clock_restart: end
[  218.434680] intel-sdw sdw-master-3: intel_resume_runtime done
[  218.932260] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  218.932343] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  220.442971] rt715 sdw:3:25d:715:0: Initialization not complete, timed out
````

I guess: there is a interrupt triggered immediately after bus reset but before interrupt is enabled. So the the interrupt for attachment is missed. I tested on CML-H. It is easy to reproduce this issue with this change, and the issue disappeared after I moved sdw_cdns_enable_interrupt ahead of bus reset.

Comment (plbossart) [reply drivers/soundwire/cadence_master.c:1398]: @RanderWang ok, I guess that makes sense, it's indeed risky if there is an interrupt generated but the interrupts are not enabled.
However we may want to make sure all interrupts status are cleared before re-enabling them prior to a bus reset?

**Comment by plbossart** [comment]
closing, fiddling with the interrupt order was a bad idea, will resubmit a new PR with only valid fixes.

File: drivers/soundwire/cadence_master.c (hunk 3 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1398]: @plbossart   yes, it is defensive programming. please check https://github.com/thesofproject/linux/pull/1671

**Pull request closed without merging**